### PR TITLE
SPEC-1396 Drivers MUST clear connection pools when SDAM monitoring fails due to a network error


### PR DESCRIPTION
**Pull request by p-mongo**

### CODE REVIEW

File: source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst [hunk 1]
```diff
@@ -1308,10 +1308,11 @@ Error handling
 Network error during server check
 `````````````````````````````````
 
-When a server `check`_ fails due to a network error,
-the client SHOULD clear its connection pool for the server:
+When a server `check`_ fails due to a network error (including a timeout),
+the client MUST clear its connection pool for the server:
```
Comment (ShaneHarvey) [inline source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: Any ideas of how we can test this? Maybe in the CMAP spec tests we can configure isMaster to fail once with a network error via configureFailPoint, wait for the monitor to run the doomed isMaster, and then assert that the pool has been cleared.

Comment (ShaneHarvey) [reply source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: I'm not saying we need tests in this PR. Just spitballing. The current changes LGTM.

Comment (p-mongo) [reply source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: I wrote several tests in https://github.com/mongodb/mongo-ruby-driver/pull/1475 but I would prefer to wait until the unified spec runner work is complete before proposing spec tests that touch both cmap and sdam.

Comment (jyemin) [reply source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: Do we need to wait for that?  It's a non-goal in the unified spec runner scope to define the format for non-”operations” tests, including CMAP and SDAM.

Comment (p-mongo) [reply source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: Our cmap spec runner does not do any i/o, thus as it is implemented it is unable to set fail points on the server.

File: source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst [hunk 2]
```diff
@@ -1308,10 +1308,11 @@ Error handling
 Network error during server check
 `````````````````````````````````
 
-When a server `check`_ fails due to a network error,
-the client SHOULD clear its connection pool for the server:
+When a server `check`_ fails due to a network error (including a timeout),
```
Comment (mbroadst) [inline source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: The spec defines terms "network error" and "network timeout", should that second term be used here?

File: source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst [hunk 3]
```diff
@@ -1308,10 +1308,11 @@ Error handling
...
+the client MUST clear its connection pool for the server:
 if the monitor's socket is bad it is likely that all are.
-(See `JAVA-1252 <https://jira.mongodb.org/browse/JAVA-1252>`_.)
+(See `JAVA-1252 <https://jira.mongodb.org/browse/JAVA-1252>`_,
+`SPEC-1396 <https://jira.mongodb.org/browse/SPEC-1396>`_.)
```
Comment (mbroadst) [inline source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: SPEC tickets are not public, I don't think we can include this in our public documentation

Comment (jyemin) [reply source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: +1

File: source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst (hunk 2 above)
Comment (p-mongo) [reply source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: Changed.

File: source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst (hunk 3 above)
Comment (p-mongo) [reply source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: Removed.

File: source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst (hunk 1 above)
Comment (p-mongo) [reply source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: @ShaneHarvey It appears that failcommand fail point is ignored for ismaster commands.

Comment (p-mongo) [reply source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst]: Turns out it does work when ismaster is given in mixed case, https://jira.mongodb.org/browse/SERVER-44414.

**Pull request merged**